File: app/src/main/java/io/github/jangerhard/BitcoinWalletTracker/utilities/SharedPreferencesHelper.java
package io.github.jangerhard.BitcoinWalletTracker.utilities;

import java.util.List;

import android.content.SharedPreferences;

public class SharedPreferencesHelper {

    private final String CURRENCY_PAIR = "currencyPair";
    private final String TOTAL_INVESTMENT = "totalInvestment";
    private final String DARK_THEME_SELECTED = "dark_theme_selected";
    private final String REFRESHING_THEME = "refreshing_theme";
    private final String SHOW_GAIN_PERCENTAGE = "show_gain_percentage";
    private final String PREFS_ACCOUNT_KEY = "BitcoinAddresses";

    private SharedPreferences preferences;

    public SharedPreferencesHelper(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    public void toggleDarkThemeSelected(boolean b) {
        preferences.edit().putBoolean(DARK_THEME_SELECTED, b).apply();
        preferences.edit().putBoolean(REFRESHING_THEME, true).apply();
    }

    public boolean isDarkTheme() {return preferences.getBoolean(DARK_THEME_SELECTED, true);}

    public boolean isThemeRefreshing() {return preferences.getBoolean(REFRESHING_THEME, false);}

    public void stopThemeRefreshing() { preferences.edit().putBoolean(REFRESHING_THEME, false).apply();}

    public String getCurrencyPair() {
        return preferences.getString(CURRENCY_PAIR, "USD");
    }

    public void saveCurrencyPair(String pair) {
        preferences.edit().putString(CURRENCY_PAIR, pair).apply();
    }

    public String getNickname(String address) {return preferences.getString(address, "Wallet");}

    public void deleteNickname(String address) { preferences.edit().remove(address).apply(); }

    public void saveNickname(String address, String name) {
        preferences.edit().putString(address, name).apply();
    }

    public long getInvestmentFromPrefs() {
        return preferences.getLong(TOTAL_INVESTMENT, 0L);
    }

    public void saveInvestment(long investment) {
        preferences.edit().putLong(TOTAL_INVESTMENT, investment).apply();
    }

    public void setRefreshingTheme(boolean b) {
        preferences.edit().putBoolean(REFRESHING_THEME, b).apply();
    }

    public boolean shouldShowGainInPercentage() {return preferences.getBoolean(DARK_THEME_SELECTED, true);}

    public void setShowGainInPercentage(boolean b) {
        preferences.edit().putBoolean(SHOW_GAIN_PERCENTAGE, b).apply();
    }

    public String getAccountsString() {
        return preferences.getString(PREFS_ACCOUNT_KEY, "");
    }

    @Deprecated
    public void saveAddresses(List<String> addresses) {
        StringBuilder addressString = new StringBuilder();
        for (String s : addresses) {
            addressString.append(s);
            addressString.append(",");
        }
        preferences.edit().putString(PREFS_ACCOUNT_KEY, addressString.toString()).apply();
    }

    public void saveTrackedWallets(io.vavr.collection.List<TrackedWallet> addresses) {
        String s = addresses
                .map(TrackedWallet::getAddress)
                .intersperse(",")
                .foldLeft(new StringBuilder(), StringBuilder::append)
                .toString();
        preferences.edit().putString(PREFS_ACCOUNT_KEY, s).apply();
    }
}

File: app/src/main/java/io/github/jangerhard/BitcoinWalletTracker/utilities/SharedPreferencesHelper.kt
package io.github.jangerhard.BitcoinWalletTracker.utilities

import android.content.SharedPreferences
import io.vavr.collection.List
import io.vavr.control.Option

class SharedPreferencesHelper(private val preferences: SharedPreferences) {

    private val CURRENCY_PAIR = "currencyPair"
    private val TOTAL_INVESTMENT = "totalInvestment"
    private val DARK_THEME_SELECTED = "dark_theme_selected"
    private val REFRESHING_THEME = "refreshing_theme"
    private val SHOW_GAIN_PERCENTAGE = "show_gain_percentage"
    private val PREFS_ACCOUNT_KEY = "BitcoinAddresses"

    fun isDarkTheme() = preferences.getBoolean(DARK_THEME_SELECTED, true)

    fun isThemeRefreshing() = preferences.getBoolean(REFRESHING_THEME, false)

    fun getCurrency(): Option<String> {

        val currency = preferences.getString(CURRENCY_PAIR, "USD")

        if (currency.isNullOrBlank()) return Option.some("USD")

        return Option.some(currency)
    }

    fun getInvestment() = preferences.getLong(TOTAL_INVESTMENT, 0L)

    fun getAddresses(): List<String> {

        val s = preferences.getString(PREFS_ACCOUNT_KEY, "")

        if (s.isNullOrBlank()) return List.empty()

        return List.ofAll(s.split(","))
    }

    fun toggleDarkThemeSelected(b: Boolean) {
        preferences.edit().putBoolean(DARK_THEME_SELECTED, b).apply()
        preferences.edit().putBoolean(REFRESHING_THEME, true).apply()
    }

    fun stopThemeRefreshing() {
        preferences.edit().putBoolean(REFRESHING_THEME, false).apply()
    }

    fun saveCurrencyPair(pair: String) {
        preferences.edit().putString(CURRENCY_PAIR, pair).apply()
    }

    fun getNickname(address: String): String? {
        return preferences.getString(address, "Wallet")
    }

    fun deleteNickname(address: String) {
        preferences.edit().remove(address).apply()
    }

    fun saveNickname(address: String, name: String) {
        preferences.edit().putString(address, name).apply()
    }

    fun saveInvestment(investment: Long) {
        preferences.edit().putLong(TOTAL_INVESTMENT, investment).apply()
    }

    fun shouldShowGainInPercentage(): Boolean =
            preferences.getBoolean(DARK_THEME_SELECTED, true)

    fun setShowGainInPercentage(b: Boolean) {
        preferences.edit().putBoolean(SHOW_GAIN_PERCENTAGE, b).apply()
    }

    fun saveTrackedWallets(addresses: List<TrackedWallet>) {
        preferences.edit().putString(PREFS_ACCOUNT_KEY,
                addresses.map { it.address }.mkString(",")).apply()
    }
}
